Implement experimental module fallback service for testing

jasnell opened the pull request:
Adds the ability to load modules lazily from a local http fallback service. The fallback service is a simple http server that returns the Worker::Module configuration for a module dynamically rather than embedded within the worker configuration. This is only usable for local dev. See samples/module_fallback for an example.

Not all modules will use the fallback service. In the worker configuration, a module has to be explicitly configured to use the fallback. Also, because of the nature of the ESM modules and the way v8 handles them, there really is no hot-module-reloading. Once a module is loaded for a worker it cannot be replaced without restarting the worker. Currently that's only possible by either shutting workerd down or using the `--watch` feature which watches for changes in the source configuration file on disk. We'll need to try work around that separately.

The module loading using the fallback service is lazy-ish. The fallback service won't be engaged until the module is actually imported but since most imports are static it will appear as if the fallback service is being accessed immediately on worker startup for most cases. You'll see the laziness of the loader in action if you're using dynamic imports or require (e.g. `import(...)` or `require(...)`). 

The fallback service is a simple http server that must return a properly JSON serialized form of the `Worker::Module` configuration as specified by the workerd.capnp schema. The request URL will be the specifier of the module in a `file://` URL form (e.g. a module specifier of `foo` will be passed to the fallback service as `file:///foo`. No other headers or details are passed in the `GET` request.

For now, every request to the fallback service spins up its own kj::Thread that is joined synchronously with the current thread. We could update that to use a single persistent thread with a queue but the current approach is simpler and raw performance really isn't a concern at this point. We can always tweak that more later.

jasnell commented on src/workerd/jsg/modules.c++:
```diff
@@ -4,6 +4,7 @@
 
 #include "jsg.h"
 #include "promise.h"
+#include "setup.h"
```
```suggestion
```

ohodson commented on samples/module_fallback/README.md:
````diff
@@ -0,0 +1,28 @@
+To run using bazel
+
+```sh
+$ bazel run //src/workerd/server:workerd -- serve ~/cloudflare/workerd/samples/helloworld_esm/config.capnp
````
```suggestion
$ bazel run //src/workerd/server:workerd -- serve ~/cloudflare/workerd/samples/module_fallback/config.capnp
```

jasnell commented:
Ok, updated to support fallback for unknown modules. Also added a header to the fallback request that indicates `import` or `require`. The example also shows that unknown built-ins like `node:vm` can be handled by the fallback service as well. The referencing specifier is not included here, instead the specifier that you get is the import specifier resolved against the reference.

jasnell commented on src/workerd/server/workerd.capnp:
```diff
@@ -254,6 +254,10 @@ struct Worker {
       # (a) allows for importing Node.js-compat built-ins without the node: specifier-prefix
       # (b) exposes the subset of common Node.js globals such as process, Buffer, etc that
       #     we implement in the workerd runtime.
+
+      fallbackService @8 :Void;
+      # A module whose type and content is determined dynamically by calling out to
+      # the configured module fallback service, if any.
```
I prefer having it for flexibility but could go either way.

kentonv commented on src/workerd/server/server.c++:
```diff
@@ -2494,6 +2497,94 @@ kj::Own<Server::Service> Server::makeWorker(kj::StringPtr name, config::Worker::
+            kj::HttpHeaderId kMethod = builder.add("resolve-method");
+            auto headerTable = builder.build();
+
+            auto addr = kj::heap<PromisedNetworkAddress>(
```
You could just `.wait(io.waitScope)` the promise here instead of wrapping it in `PromisedNetworkAddress`.

kentonv commented on src/workerd/server/server.c++:
```diff
@@ -2494,6 +2497,94 @@ kj::Own<Server::Service> Server::makeWorker(kj::StringPtr name, config::Worker::
+
+            auto request = client->request(
+                kj::HttpMethod::GET,
+                kj::str("file://", specifier),
```
An HTTP request path should start with `/`. A protocol prefix like `file://` is not expected here.

This code as-is will produce an HTTP request line like:

```
GET file://blah HTTP/1.1
```

jasnell replied:
Yes, that's intentional here. The target identifier portion of the HTTP request line is a `URI Reference` which can be a fully-qualified absolute URL.

kentonv replied:
FWIW the HTTP spec says you "MUST" use origin form (path, not URL) when talking to an origin server, and "MUST" use absolute form when talking to a proxy.

https://www.ietf.org/archive/id/draft-ietf-httpbis-messaging-11.html#name-origin-form

Admittedly I generally feel that if the client and server want to agree on something else, they are free to do so, regardless of what MUSTs appear in the spec. But it does seem clear that the HTTP spec does not intend for full URLs to be used in this context.

kentonv replied:
The paragraphs you quote are interesting and I didn't realize origin servers were expected to support this. However, these pagagraphs seem to very clearly imply that the request-target should be an HTTP URL and that the hostname within that URL should be the hostname of the origin server. Sending `file://` seems clearly not in the spirit of what they're saying here.

I think the only way I could be convinced that this is an appropriate way to use the HTTP protocol is if you really go all in on arguing that this is a proxy. But even then, this is a very... unusual design. I would guess that a lot of HTTP server frameworks won't even support these kinds of requests. I suppose since it's a private protocol, the only HTTP server that matters is whatever miniflare uses. But it feels like asking for trouble.

kentonv commented on src/workerd/server/server.c++:
```diff
@@ -2494,6 +2497,94 @@ kj::Own<Server::Service> Server::makeWorker(kj::StringPtr name, config::Worker::
+            kj::AsyncIoContext io = kj::setupAsyncIo();
+
+            kj::HttpHeaderTable::Builder builder;
+            kj::HttpHeaderId kMethod = builder.add("resolve-method");
```
Nitpicks:

* In KJ code we normally capitalize this like `Resolve-Method`, though it doesn't really matter.
* Since this isn't a standard header it should probably have an `X-` prefix?

jasnell replied:
@mrbbot ... just a heads up, I went ahead and renamed the header to `x-resolve-method`.

kentonv commented on src/workerd/server/server.c++:
```diff
@@ -2494,6 +2497,94 @@ kj::Own<Server::Service> Server::makeWorker(kj::StringPtr name, config::Worker::
+              auto payload = resp.body->readAllText().wait(io.waitScope);
+              KJ_LOG(ERROR, "Fallback service failed to fetch module", payload);
+            } else {
+              jsonPayload = kj::str(resp.body->readAllText().wait(io.waitScope));
```
`kj::str()` is redundant, `readAllText().wait()` already returns `kj::String`.

jasnell commented:
@mrbbot ...

> Returning a module from the fallback service with a different name than the requested path doesn't really work...

This becomes a bit tricky to implement but not impossible. It would largely mean that an individual `ModuleInfo` could be indexed multiple times in the entries table with different names. Not a huge refactor but I think the changes needed for this would also fix the reallocation issue so definitely worth doing.

> Requests to the module fallback service don't differentiate between the referencing path and specifier, they only give you the URL of the module to load. 

Just to make sure I understand, you'd like the request to the fallback service to include both the specifier of the module being imported/required and the specifier of the referent? Should be possible with a few additional minor tweaks.

> For example, importing node:assert in a/b/c/index.mjs will try fetch /a/b/c/node:assert from the fallback service, then if that fails try to load /node:assert which succeeds

This one should be fixed with the refactor that treats specifiers as proper URLs but I'll verify that. This is a bit of a larger change so I might not do anything for this one immediately and just handle it as part of the larger refactor.

> Circular require() is not supported. Modules like chai try to require the root of their module in submodules

Yeah, this is one that I've been wanting to revisit for a while. For now since you have a workaround let's leave it as is. After the larger refactor is done I'll revisit circular requires/imports in general.

jasnell commented on src/workerd/server/server.c++:
```diff
@@ -2494,6 +2497,94 @@ kj::Own<Server::Service> Server::makeWorker(kj::StringPtr name, config::Worker::
+            } else {
+              jsonPayload = kj::str(resp.body->readAllText().wait(io.waitScope));
+            }
+          } catch (...) {
```
Yeah, I know. I have a general preference for keeping the try/catch within the child thread here, especially since we're not actually propagating the error up but simply logging it.

jasnell commented:
ok, I'll likely be able to work up changes for this by end of day tomorrow.

jasnell commented on src/workerd/server/server.c++:
```diff
@@ -2494,6 +2497,152 @@ kj::Own<Server::Service> Server::makeWorker(kj::StringPtr name, config::Worker::
+        }
+      }
+      if (!prefixed) {
+        actualSpecifier = specifier;
```
I applied this check a bit further down. Please take a look and see if it's behaving correctly now.

jasnell commented on src/workerd/jsg/modules.c++:
ok, I have it now checking to see if the `resolve()` call has the `BUILTIN_ONLY` option and if it does, if the specifier is prefixed with `node:`, `cloudflare:`, or `workerd:` it will apply `Type::BUILTIN` to the added module.

jasnell commented:
Ok, internal CI run looks good and shows this can land without requiring internal changes to land at the same time. Will get this merged.
